Components/Queues/Handler/PhpAmqpHandler.php

```diff
@@ -0,0 +1,232 @@
+     * Dispatch the command to stop the work on next iteration
+     * @param string $consumerTag
+     * @throws \Exception
+     */
+    public function stopConsume(string $consumerTag)
+    {
+        try {
+            $this->stop();
+            if ($this->channel && $this->channel->getConnection()) {
+                $this->channel->basic_cancel($consumerTag, true, true);
+            }
+        } catch (\Exception $exception) {
+            $this->log($exception->getMessage());
+            exit(1);
+        }
+    }
+
+    /**
+     * Verifies if the message object is an instance of QueueMesseageInterface
+     * and if the handler is not an instance of QueueMessageHandlerInterface. Used by dispatchMessage function
+     * @param MessageInterface $message
+     * @return bool
+     */
+    private function isQueueMessage($message): bool
+    {
+        return $message instanceof QueueMessageInterface && !($this->endpoint->getHandler() instanceof QueueMessageHandlerInterface);
+    }
+
+    /**
+     * Log in console the message passed as param
+     * @param string $message
+     * @param array $ctx
+     * @throws \Exception
+     */
+    private function log(string $message, array $ctx = [])
+    {
+        if (null === $this->logger) {
+            return;
+        }
+
+        $now = new \DateTime();
+        $ctx['time'] = $now->format('Y-m-d H:i:s.u');
+        $this->logger->info($message, $ctx);
+    }
+
+    /**
+     * Dispatch the message to target system or events to database
+     * @param QueueMessage $message
+     * @throws \Exception
+     */
+    private function dispatchMessage(QueueMessage $message)
+    {
+        if (!$this->endpoint) {
+            $this->getExceptionHandler()('Endpoint is undefined');
+            return;
+        }
+
+        try {
+            if ($this->isQueueMessage($message) && null !== $this->smartesbHelper) {
+                $endpoint = $this->smartesbHelper->getEndpointFactory()->createEndpoint($message->getDestinationURI(), EndpointFactory::MODE_CONSUME);
+                $endpoint->getHandler()->handle($message->getBody(), $endpoint);
+            } else {
+                $this->endpoint->handle($message);
+            }
+        } catch (\Exception $exception) {
+            $this->getExceptionHandler()($exception, ['headers' => $message->getHeaders(), 'body' => $message->getBody()]);
+        }
+    }
+
+    /**
+     * Deserialize the messages and dispatch them to the other targets
+     * @param AMQPMessage
+     * @return  MessageInterface|QueueMessageInterface $message
+     */
+    public function deserializeMessage(AMQPMessage $message): QueueMessageInterface
+    {
+        try {
+            return $this->serializer->deserialize($message->getBody(), SerializableInterface::class, $this->format);
+        } catch (\Exception $exception) {
+            $this->getExceptionHandler()($exception, ['headers' => $message->getHeaders(), 'body' => $message->getBody()]);
```

Comment:
FYI, this is the only correct usage of the ExceptionHandler, every other call to it should be removed.

Reply:
Alright, see the comment above.